File: src/main/java/frc/robot/subsystems/intake/IntakeIOSimBasic.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.subsystems.intake;

import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.trajectory.TrapezoidProfile;
import frc.robot.Constants.IntakeConstants;
import frc.robot.Constants.SimConstants;

/** Add your docs here. */
public class IntakeIOSimBasic implements IntakeIO {
    private TrapezoidProfile.State state = new TrapezoidProfile.State(IntakeConstants.STOWED.getRotations(), 0);
    private double rollerSpeed = 0.0;
    private TrapezoidProfile profile = new TrapezoidProfile(
            new TrapezoidProfile.Constraints(IntakeConstants.CRUISE_VELOCITY, IntakeConstants.MAX_ACCELERATION));
    private TrapezoidProfile.State goal = state;

    public void stopArmMotor() {
        goal = new TrapezoidProfile.State(state.position, 0);
    }

    public void setRollerSpeed(double speed) {
        rollerSpeed = 20 * speed;
    }

    public void stopRollerMotor() {
        rollerSpeed = 0.0;
    }

    public void setRollerVoltage(double voltage) {
        setRollerSpeed(voltage / 12.0);
    }

    public void setArmPosition(Rotation2d position) {
        goal = new TrapezoidProfile.State(position.getRotations(), 0);
    }

    public void updateInputs(IntakeIOInputs inputs) {
        state = profile.calculate(SimConstants.LOOP_TIME, state, goal);
        inputs.armPosition = Rotation2d.fromRotations(state.position);
        inputs.armVelocity = state.velocity;
        // trick homing code
        if (inputs.armPosition.equals(IntakeConstants.STOWED)) {
            inputs.armCurrentAmps = IntakeConstants.ARM_DEPLOY_CURRENT_THRESHOLD + 1;
        } else {
            inputs.armCurrentAmps = 0;
        }
        inputs.rollerVelocity = rollerSpeed;
    }
}

File: src/main/java/frc/robot/subsystems/intake/IntakeIOSimBasic.kt
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.
package frc.robot.subsystems.intake

import com.lobstahbots.units.*
import edu.wpi.first.math.geometry.Rotation2d
import edu.wpi.first.math.trajectory.TrapezoidProfile
import frc.robot.Constants.IntakeConstants
import frc.robot.Constants.SimConstants
import frc.robot.subsystems.intake.IntakeIO.IntakeIOInputs

/** Add your docs here.  */
class IntakeIOSimBasic : IntakeIO {
    private var state = TrapezoidProfile.State(IntakeConstants.STOWED.rotations, 0.0)
    private var rollerSpeed = 0.0
    private val profile = TrapezoidProfile(
        TrapezoidProfile.Constraints(IntakeConstants.CRUISE_VELOCITY, IntakeConstants.MAX_ACCELERATION)
    )
    private var goal: TrapezoidProfile.State = state

    override fun stopArmMotor() {
        goal = TrapezoidProfile.State(state.position, 0.0)
    }

    override fun setRollerSpeed(speed: Double) {
        rollerSpeed = 20 * speed
    }

    override fun stopRollerMotor() {
        rollerSpeed = 0.0
    }

    override fun setRollerVoltage(volts: Double) {
        setRollerSpeed(volts / 12.0)
    }

    override fun setArmPosition(position: Rotation2d) {
        goal = TrapezoidProfile.State(position.rotations, 0.0)
    }

    override fun updateInputs(inputs: IntakeIOInputs) {
        state = profile.calculate(SimConstants.LOOP_TIME, state, goal)
        inputs.armPosition = Rotation2d.fromRotations(state.position)
        inputs.armVelocity = state.velocity.rotationsPerSecond
        // trick homing code
        if (inputs.armPosition == IntakeConstants.STOWED) {
            inputs.armCurrent = IntakeConstants.ARM_DEPLOY_CURRENT_THRESHOLD.amps + 1.amp
        } else {
            inputs.armCurrent = 0.amps
        }
        inputs.rollerVelocity = rollerSpeed.rotationsPerSecond
    }
}
